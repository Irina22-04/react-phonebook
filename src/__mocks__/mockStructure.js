export const mockStructure = [
    {
        "name": "OOO 'Дети дикой обезьяны'",
        "id": 1
    },
    {
        "name": "Управление 1",
        "parent": 1,
        "id": 2
    },
    {
        "name": "Управление 22",
        "parent": 1,
        "id": 3
    },
    {
        "name": "Управление 32",
        "parent": 1,
        "id": 4
    },
    {
        "name": "Отдел по работе с клиентами",
        "id": 5,
        "parent": 2
    },
    {
        "name": "Отдел по работе с юридическими лицами",
        "id": 8,
        "parent": 2
    }
];