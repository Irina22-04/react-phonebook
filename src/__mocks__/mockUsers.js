export const mockUsers = [
    {
        "name": "Дмитрий",
        "surname": "Дмитриевич",
        "patronym": "Дмитриевич",
        "department": 8,
        "fullname": "Дмитриев Дмитрий Дмитриевич",
        "id": 104,
        "isAdmin": true,
        "email": "admin@mail.ru",
        "password": "$2a$10$AnPxLMUGPKB8eAsdF90pFOWi/h2ejtPhrkzA6.a/nNUty8SexQBQe",
        "position": "специалист",
        "phone": "340-34-34",
        "businessMobile": "(33)333-33-33",
        "personalMobile": "(44)444-44-44",
        "personalEmail": "Dmit@mail.ru",
        "birthday": "22/02/1999"
    },
    {
        "name": "Михаил",
        "surname": "Михайлов",
        "patronym": "Михайлович",
        "department": 14,
        "fullname": "Михайлов Михаил Михайлович",
        "id": 109,
        "position": "специалист",
        "phone": "340-34-34"
    }
];