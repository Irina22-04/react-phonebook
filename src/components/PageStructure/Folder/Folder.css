.folder {
    font-size: 18px;
    display: inline-block;
    margin-left: 30px;
    width: 100%;
    max-width: 500px;
}

.sign {
    display: inline-block;
    margin-left: -10px;
}

.folders-wrapper {
    margin-left: 30px;
}

.set-department {
    font-weight: bold;
}

.name-wrapper {
    display: flex;
    align-items: center;
    margin-left: 6px;
}

.sign {
    width: 20px;
    height: 20px;
    border: 1px solid black;
    border-radius: 10px;
    display: flex;
    align-items: center;
    justify-content: center;
}

.item-enter {
    transform: translateX(-50%);
}

.item-enter-active {
    transform: translateX(0);
    transition: transform 200ms ease-in;
}

.item-exit {
    transform: translateX(0);
}

.item-exit-active {
    transform: translateX(-50%);
    transition: transform 200ms ease-in;
}

.structure > .folder:after {
    display: none
}

.structure > .folder:before {
    display: none
}

@media (max-width: 1100px) {
    .name-wrapper {
        max-width: 300px;
    }
}

@media (max-width: 720px) {
    .folder {
        width: 100%;
        min-width: 320px;
        max-width: 400px;
    }

    .folders-wrapper {
        margin-left: 15px;
    }

    .folder {
        margin-left: 15px;
    }
}

@media (max-width: 480px) {
    .folder {
        font-size: 16px;
        width: 100%;
        margin-left: 10px;
        min-width: 200px;
        max-width: 200px;
    }

    .folders-wrapper {
        margin-left: 10px;
    }
}

