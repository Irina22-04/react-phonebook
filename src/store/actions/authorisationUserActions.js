import {get} from "lodash";
import jwt_decode from "jwt-decode";

import {getContactById, loginUser} from "../../utils/api";

export const USER_AUTHORISATION = 'USER_AUTHORISATION';
export const LOGIN_TRY = 'LOGIN_TRY';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAIL = 'LOGIN_FAIL';
export const LEAVE_ACCOUNT = 'LEAVE_ACCOUNT';

export function setUserAuthorisation(dispatch, authorisationState) {
    return dispatch({
        type: USER_AUTHORISATION,
        payload: authorisationState,
    })
}

export const loginUserTry = (dispatch, user) => async (dispatch) => {
    dispatch({type: LOGIN_TRY});
    return await loginUser(user)
        .then(({data}) => {
            const {sub} = jwt_decode(data.accessToken);
            return sub
        })
        .then(async (id) => await getContactById(id))
        .then(({data}) => {
            dispatch({
                type: LOGIN_SUCCESS,
                payload: data,
            })
        })
        .catch(e => {
            dispatch({
                type: LOGIN_FAIL,
                payload: get(e, ['response', 'data', 'error_message'], e.message)
            });
            return Promise.reject('Ошибка авторизации')
        });
};

export function leaveAccount(dispatch) {
    return dispatch({
        type: LEAVE_ACCOUNT,
    })
}