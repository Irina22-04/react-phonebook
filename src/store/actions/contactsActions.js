import {get} from "lodash";

import {getAllContacts} from "../../utils/api";

export const CONTACTS_LOADING = 'CONTACTS_LOADING';
export const CONTACTS_SET = 'CONTACTS_SET';
export const CONTACTS_ERROR = 'CONTACTS_ERROR';
export const CONTACTS_SET_PAGE = 'CONTACTS_SET_PAGE';
export const CONTACTS_SET_SURNAME_LETTER = 'CONTACTS_SET_SURNAME_LETTER';
export const CONTACTS_SET_FULLNAME_PATTERN = 'CONTACTS_SET_FULLNAME_PATTERN';
export const RESET_CONTACT_SEARCH = 'RESET_CONTACT_SEARCH';


export function getAllContactsTry(dispatch) {
    return async function () {
        dispatch({type: CONTACTS_LOADING});
        await getAllContacts()
            .then(data => {
                dispatch({
                    type: CONTACTS_SET,
                    payload: data.data,
                })
            })
            .catch(e => {
                dispatch({
                    type: CONTACTS_ERROR,
                    payload: get(e, ['response', 'data', 'error_message'], e.message)
                });
            });
    }
}

export function setPage(dispatch, page) {
    return dispatch({type: CONTACTS_SET_PAGE, payload: page});
}

export function setSurnameLetter(dispatch, letter) {
    return dispatch({type: CONTACTS_SET_SURNAME_LETTER, payload: letter});
}

export function setFullnamePattern(dispatch, pattern) {
    return dispatch({type: CONTACTS_SET_FULLNAME_PATTERN, payload: pattern});
}

export function resetSearchParams(dispatch) {
    return dispatch({type: RESET_CONTACT_SEARCH});
}
