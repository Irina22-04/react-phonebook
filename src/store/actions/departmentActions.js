import {get} from "lodash";

import {addStructure, deleteStructure, getStructure, updateStructure} from "../../utils/api";

export const STRUCTURE_LOADING = 'STRUCTURE_LOADING';
export const STRUCTURE_SET = 'STRUCTURE_SET';
export const STRUCTURE_ERROR = 'STRUCTURE_ERROR';
export const EDIT_DEPARTMENT = 'SET_EDIT_DEPARTMENT';
export const SET_DEPARTMENT = 'SET_DEPARTMENT';
export const STRUCTURE_UPDATE = 'STRUCTURE_UPDATE';
export const STRUCTURE_DELETE = 'STRUCTURE_DELETE';
export const SET_PARENT_DEPARTMENT = 'SET_PARENT_DEPARTMENT';
export const STRUCTURE_ADD = 'STRUCTURE_ADD';
export const CARRY_DEPARTMENT = 'CARRY_DEPARTMENT';
export const SET_DELETE_DEPARTMENT_ERROR = 'SET_DELETE_DEPARTMENT_ERROR';
export const REMOVE_DELETE_DEPARTMENT_ERROR = 'REMOVE_DELETE_DEPARTMENT_ERROR';

export function setDepartment(dispatch, department) {
    return dispatch({
        type: SET_DEPARTMENT,
        payload: department,
    })

}

export function editDepartment(dispatch, department) {
    return dispatch({
        type: EDIT_DEPARTMENT,
        payload: department,
    })

}

export function getStructureTry(dispatch) {
    return async function () {
        dispatch({type: STRUCTURE_LOADING});
        await getStructure()
            .then(data => {
                dispatch({
                    type: STRUCTURE_SET,
                    payload: data.data,
                })
            })
            .catch(e => {
                dispatch({
                    type: STRUCTURE_ERROR,
                    payload: get(e, ['response', 'data', 'error_message'], e.message)
                });
            });
    }
}

export function updateStructureTry(dispatch, department) {
    return async function () {
        await updateStructure(department)
            .then(data => {
                dispatch({
                    type: STRUCTURE_UPDATE,
                    payload: data.data,
                })
            })
            .catch(e => {
                return Promise.reject(e.message);
            });
    }
}

export function deleteStructureTry(dispatch, department) {
    return async function () {
        await deleteStructure(department)
            .then(() => {
                dispatch({
                    type: STRUCTURE_DELETE,
                    payload: department,
                })
            })
            .catch(e => {
                return Promise.reject(e.message);
            });
    }
}

export function addChildrenDepartment(dispatch, parentDepartment) {
    return dispatch({
        type: SET_PARENT_DEPARTMENT,
        payload: parentDepartment,
    })

}

export function addStructureTry(dispatch, department) {
    return async function () {
        await addStructure(department)
            .then((data) => {
                dispatch({
                    type: STRUCTURE_ADD,
                    payload: data.data,
                });
            })
            .catch(e => {
                return Promise.reject(e.message);
            });
    }
}

export function carryDepartment(dispatch, department) {
    return dispatch({
        type: CARRY_DEPARTMENT,
        payload: department,
    })
}

export function deleteDepartmentError(dispatch, error) {
    return dispatch({
        type: SET_DELETE_DEPARTMENT_ERROR,
        payload: error,
    })
}

export function removeDeleteDepartmentError(dispatch) {
    return dispatch({
        type: REMOVE_DELETE_DEPARTMENT_ERROR,
    })
}
