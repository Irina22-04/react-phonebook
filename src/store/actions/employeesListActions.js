import {get} from "lodash";

import {getEmployeesByStructure} from "../../utils/api";

export const EMPLOYEES_LOADING = 'EMPLOYEES_LOADING';
export const EMPLOYEES_SET = 'EMPLOYEES_SET';
export const EMPLOYEES_ERROR = 'EMPLOYEES_ERROR';

export function getEmployeesByStructureTry(dispatch, department) {
    return async function () {
        dispatch({type: EMPLOYEES_LOADING});
        await getEmployeesByStructure(department)
            .then(data => {
                dispatch({
                    type: EMPLOYEES_SET,
                    payload: data.data,
                })
            })
            .catch(e => {
                dispatch({
                    type: EMPLOYEES_ERROR,
                    payload: get(e, ['response', 'data', 'error_message'], e.message)
                });
            });
    }
}
