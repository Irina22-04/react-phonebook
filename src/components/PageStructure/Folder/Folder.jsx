import React from "react";
import {connect} from "react-redux";
import {CSSTransition, TransitionGroup} from "react-transition-group";
import PropTypes from "prop-types";

import {getEmployeesByStructureTry} from "../../../store/actions/employeesListActions";
import {setDepartment} from "../../../store/actions/departmentActions";
import "./Folder.css";

class prevFolder extends React.PureComponent {

    static propTypes = {
        employeesListForDepartment: PropTypes.object.isRequired,
        department: PropTypes.object.isRequired,
        folderId: PropTypes.number.isRequired,
    };

    constructor(props) {
        super(props);
        this.state = {
            isSelected: false,
            children: null,
        }
    }

    componentDidMount() {
        if (this.props.department.status === 3) {
            this.setState({
                children: this.props.department.structure.filter(department => department.parent === this.props.folderId),
            })
        }
    }

    checkArrow = () => {
        const {children} = this.state;

        const isChildren = children && children.length;
        return isChildren ? this.createSign() : null;
    };

    createSign = () => {
        const {isSelected} = this.state;
        const sign = isSelected ? '-' : '+';
        return <div className={'sign'}>{sign}</div>
    };

    switchSelect = (e) => {

        const {isSelected} = this.state;
        this.setState({
            isSelected: !isSelected,
        });

        let department = e.currentTarget.getAttribute('id');

        if (department) {
            this.getEmployeesByDepartment(department);
            this.setDepartment(department);
        }

    };

    setDepartment = (department) => {
        this.props.dispatch(setDepartment(this.props.dispatch, department));
    };

    getEmployeesByDepartment = (department) => {
        this.props.dispatch(getEmployeesByStructureTry(this.props.dispatch, department));
    };

    showChildren = () => {

        const {isSelected} = this.state;
        if (!isSelected) {
            return null;
        }
        const {children} = this.state;
        if (!children || children.length === 0) {
            return null;
        }

        return children.map(child => (
            <CSSTransition timeout={200} classNames="item" key={child.name}>
                <Folder folderId={child.id}/>
            </CSSTransition>)
        );
    };

    render() {
        const department = Number(this.props.department.department);
        const departmentName = this.props.department.structure ? this.props.department.structure.find(item => item.id === this.props.folderId) : null;

        return this.props.department.status <= 1 ? null : (
            <div className={"department-wrapper folder"}>
                <div
                    id={this.props.folderId}
                    className={department === this.props.folderId ? "set-department name-wrapper" : "name-wrapper"}
                    onClick={this.switchSelect}>
                    {this.checkArrow()}
                    {departmentName.name}
                </div>
                <TransitionGroup className={"folders-wrapper"}>
                    {this.showChildren()}
                </TransitionGroup>
            </div>
        )
    }
}

function mapStateToProps(state) {
    return {
        employeesListForDepartment: state.employeesListForDepartment,
        department: state.department,
    };
}

const Folder = connect(mapStateToProps)(prevFolder);

export default Folder;