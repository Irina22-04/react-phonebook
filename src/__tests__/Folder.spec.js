import React from "react";
import renderer, {act} from "react-test-renderer";
import {Provider} from "react-redux";
import {mockStructure} from "../__mocks__/mockStructure";
import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";


import Folder from "../components/PageStructure/Folder/Folder";
import {EMPLOYEES_LOADING} from "../store/actions/employeesListActions";


const initialState = {
    department: {
        department: '',
        structure: mockStructure,
        status: 3,
        error: '',
        editDepartment: '',
        parentDepartment: '',
        carriedDepartment: '',
    },
    employeesListForDepartment: {
        status: 0,
        employeesList: [],
        error: '',
    }
};

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);


describe('PageStructure', () => {
    const state = {...initialState};
    const store = mockStore(() => state);
    const component = renderer.create(<Provider store={store}><Folder folderId={4}/></Provider>);

    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();

    it('onClick should work', () => {
        const folder = component.root.findByProps({id: 4});
        folder.props.onClick({currentTarget: {getAttribute: () => 4}});

        const expectedActions = [
            {type: EMPLOYEES_LOADING},
            {type: "SET_DEPARTMENT", payload: 4},
            {type: "SET_DEPARTMENT", payload: 4}
        ];

        expect(store.getActions()).toEqual(expectedActions);

        let componentTree = component.toJSON();
        expect(componentTree).toMatchSnapshot();
    });

});