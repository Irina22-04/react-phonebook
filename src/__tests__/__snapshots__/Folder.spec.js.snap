// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="department-wrapper folder"
>
  <div
    className="name-wrapper"
    id={4}
    onClick={[Function]}
  >
    Управление 32
  </div>
  <div
    className="folders-wrapper"
  />
</div>
`;

exports[`PageStructure onClick should work 1`] = `
<div
  className="department-wrapper folder"
>
  <div
    className="name-wrapper"
    id={4}
    onClick={[Function]}
  >
    Управление 32
  </div>
  <div
    className="folders-wrapper"
  />
</div>
`;
